### ISSUE TITLE
logstash 1.5.0 freeze

### ISSUE THREAD

**Issue by KlavsKlavsen** [elastic/logstash#3440]
After upgrading to logstash 1.5.0 (from 1.4.2), it freezes atleast once - every day (so far - has gone on for 5 days now).. 

The logstash instance(s) where this happens (all 4 of them), have 1 job - input from redis and output to ES.

When I try to stop it - it fails to stop as well:
 /etc/init.d/logstash stop
Killing logstash (pid 8979) with SIGTERM
Waiting logstash (pid 8979) to die...
Waiting logstash (pid 8979) to die...
Waiting logstash (pid 8979) to die...
Waiting logstash (pid 8979) to die...
Waiting logstash (pid 8979) to die...
logstash stop failed; still running.

from my graphs (of es indexing rate and redis queue size) it seems that it especially happens around "index rollover" where the redis queue builds up, and then after 5-10 minutes, the indexing rate goes to 0..

How can I debug this further?

**Comment by KlavsKlavsen** [comment]
the only thing logstash logs is when redis app. can't be connected to:
{:timestamp=>"2015-06-15T08:27:51.456000+0200", :message=>"Redis connection problem", :exception=>#<Redis::TimeoutError: Connection timed out>, :level=>:warn}

but even though it logs this, the logstash init script can't stop logstash.

**Comment by dforste** [comment]
I am having a similar issue with an almost identical setup from what I can tell. I was able to get a thread dump of the unresponsive processes with jstack -F 
https://gist.github.com/dforste/660059a12d2b82206776

RHEL 6.6 
LS 1.5.0
Java jdk1.8.0_31

**Comment by KlavsKlavsen** [comment]
It seems to have helped the issue, that I lowered the redis max queue size to 1GB.. atleast it hasn't happened since. 

I inserted a statsd section (setting sender => "myhostname") - to get a counter on what logstash is doing.. so I can see when anyone of the LS processes freezes.. it seems 2 (out of 4) have frozen currently.. I'll restart them later and see if they "start again"

**Comment by dforste** [comment]
Per comments in https://github.com/elastic/logstash/issues/2894 I am trying to change output to http to see if that helps.

**Comment by ph** [comment]
Are you using the multiline filter or metrics filter in your configuration, I don't see anything mentioning it but we have just released 1.5.1 (today) that include a fix for a deadlock issue with theses plugins?

**Comment by KlavsKlavsen** [comment]
I was running es http and this happened.. and then I swithed to "default" transport instead (to ingest faster)..

**Comment by ph** [comment]
@KlavsKlavsen Would you mind posting your configuration that hang logstash?

**Comment by jordansissel** [comment]
>then I swithed to "default" transport instead (to ingest faster)..

If http is slower than the default transport, then there's a bug.

**Comment by KlavsKlavsen** [comment]
input {
  redis {
        host => "p-elasticlog.internal.idk"
        # Remember that type does NOT overwrite trype from shipper!
    type => "redis-input"
    # these settings should match the output of the agent
    data_type => "list"
    key => "logstash"
    codec => json
    threads => 8
  }
  redis {
        host => "p-elasticlog.internal.idk"
        type => "netflow"
    data_type => "list"
    key => "pet1year"
    codec => json
    threads => 2
  }
  redis {
        host => "p-elasticlog01.internal.idk"
        # Remember that type does NOT overwrite trype from shipper!
    type => "redis-input"
    # these settings should match the output of the agent
    data_type => "list"
    key => "logstash"
    codec => json
    threads => 8
  }
  redis {
        host => "p-elasticlog01.internal.idk"
        type => "netflow"
    data_type => "list"
    key => "pet1year"
    codec => json
    threads => 2
  }
}

filter {
  #choose index
  if [type] == "cnrdhcp" {
    mutate { add_field => { "[@metadata][index]" => "cnrdhcp-%{+YYYY.MM.dd}" } }
  } else if [type] == "netflow" {
    mutate { add_field => { "[@metadata][index]" => "netflow-%{+YYYY.MM.dd}" } }
  } else if [type] == "cdn_access_logs" {
    mutate { add_field => { "[@metadata][index]" => "cdn_access_logs-%{+YYYY.MM.dd}" } }
  } else if [type] == "mpf_arkiv" {
    mutate { add_field => { "[@metadata][index]" => "mpf_arkiv-%{+YYYY.MM}" } }
  } else {
    mutate { add_field => { "[@metadata][index]" => "logstash-%{+YYYY.MM.dd}" } }
  }
  #generate message_id if its not present..
  if [message_id] {
    mutate { add_field => { "[@metadata][id]" => "%{message_id}" } }
    mutate { remove_field => message_id }
    mutate { add_tag => "hasmessage_id" }
  } else if [message] {
    ruby {
      init => "require 'digest/sha1'"
      code => "event['@metadata']['id'] = Digest::SHA1.base64digest(event['message'])"
    }
  } else {
    #really broken input.. use timestamp as id for now..  - we should never land here
    mutate { add_field => { "[@metadata][id]" => "%{@timestamp}" } }
  }
}


output {
    elasticsearch {
      codec => plain {
         charset => 'UTF-8'
      }
      host => "127.0.0.1"
      index => "%{[@metadata][index]}"
      manage_template => false
#      protocol => http
#      port => 9200
      document_id => "%{[@metadata][id]}"
    }
  statsd {
    host => "localhost"
    port => 8125
    sender => "p-elasticlog04"
    namespace => "servers"
    increment => "logstash.processing"
  }

**Comment by colinsurprenant** [comment]
the original problem definitely seems related to the deadlock issue fixed in #3378 described in issue #3361

**Comment by colinsurprenant** [comment]
I forgot to add that per the original description «it seems that it especially happens around "index rollover" where the redis queue builds up»

**Comment by KlavsKlavsen** [comment]
in regards to speed 1.5.0 seems a lot slower.. with 1.4.2 (or perhaps earlier) with default transport, I could get upto 1.5mill. docs/min (when queue had built up -normal avg. is 100-200k docs/min) - when all 4 was running, but now I only get ~300k with two working. Right now, only 2 of the logstash agents reports numbers (via statsd). counters from that is ~1800k/min from both - and the other two.. one reports 0, and the other does not report at all ( I just restarted again to see if I can revive it).

**Comment by KlavsKlavsen** [comment]
with http transport with all 4 (before 1.5.0) I could get a max of about 600k docs/min.

**Comment by KlavsKlavsen** [comment]
I have numbers from ES (index size etc.) collected every minute for the last year.. so I have a lot of numbers if you want them.. and all changes to production (except manual restarts of services) is in version control ofcourse.. so I can pinpoint when changes has been commited :)

**Comment by colinsurprenant** [comment]
@KlavsKlavsen could you test with 1.5.1? 

Looking at the 1.5.0 deadlock code we fixed, I think it might be the problem here but there is also an unsolved shutdown problem where if the output plugins are completely stalled, the shutdown event will not be correctly propagated until the output plugins are un-stalled.

**Comment by ph** [comment]
@KlavsKlavsen which version of the `logstash-output-elasticsearch` you are currently using? You can get the version by running this command `bin/plugin list --verbose logstash-output-elasticsearch`. We have changed the defaults in the es output related to timeout that could probably help.  

If it is slower we certainly need to fix this!

**Comment by colinsurprenant** [comment]
@KlavsKlavsen I suggest we move the performance discussion in another issue and keep this one about the original shutdown problem. When you create that performance issue could you please explain/clarify the performance numbers you mentioned above - I have a real hard time understanding what you reported. maybe format it table-like?

**Comment by KlavsKlavsen** [comment]
certainly - I actually wasn't trying to start a performance discussion.. it's adaquate as it is, and I expect it to scale, if I add more ES nodes anyways..

**Comment by KlavsKlavsen** [comment]
Attached graph of how logstash "stops processing" and starts again.. or completely drops out.
![es-graph](https://cloud.githubusercontent.com/assets/2737885/8200714/5bbee806-14c9-11e5-9589-f968e6408bfa.png) (see logstash statsd output in lower right corner). ES Indexing rate is calculated by taking the diff in doc count rfrom ES.

**Comment by KlavsKlavsen** [comment]
and when 02 (the green line in lower right corner) started processing, was after I killed and started logstash.

**Comment by KlavsKlavsen** [comment]
# /opt/logstash/bin/plugin list --verbose logstash-output-elasticsearch
logstash-output-elasticsearch (0.2.4)
logstash-output-elasticsearch_http (0.0.2)

**Comment by KlavsKlavsen** [comment]
the shutdown problem should also be handled in init script.. instead of simply failing to stop - it should end up KILL'ing logstash.. not just giving up.

**Comment by KlavsKlavsen** [comment]
and btw. if anyone can explain why numbers from statsd is not similar to indexing rate.. I'd be very interested..  :)

**Comment by colinsurprenant** [comment]
@KlavsKlavsen here's what I suggest: upgrade to logstash 1.5.1 & make sure you have the latest logstash-output-elasticsearch, 0.2.8 is the latest as of now. You can update with `bin/plugin update logstash-output-elasticsearch`. We did fix an important "logstash shutdown freeze" condition in 1.5.1 but there is still other conditions where it can happen as described in #3451 

Also, if possible let's try to discuss one problem/subject per issue. Do not hesitate to use [IRC](https://webchat.freenode.net/?channels=logstash) or [Discuss](https://discuss.elastic.co/c/logstash) for any questions/discussions or create a separate issue for a difference problem/concern/subject.

**Comment by splitice** [comment]
+1 I can confirm this issue too. Got one worker which I test upgraded from 1.4.x. 

Neither processes anything, or stops when asked to stop.

Nothing in the logs to indicate a problem, just frozen.

All plugins up to date, and core at version 1.5.1

**Comment by KlavsKlavsen** [comment]
I've currently downgraded all to 1.4.2 to confirm that it is then stable again. After that - I'll upgrade 50% to 1.5.1 and see what happens. also - sometimes I've had to restart redis as well.. so it seems to be between logstash and redis - something is going awry.

**Comment by dirtdiver512** [comment]
All,

Seems like I am having the exact same issue.  I have a super basic config with a very low EPS, maybe 100.  It seem sLS just stops/hangs.  I can not stop the service and need to kil it with a -9.  Nothing in the logs.  I turned up verbose logging and only have the below in last line:

logstash 1.5.0
logstash-output-elasticsearch (0.2.4)
logstash-output-elasticsearch_http (0.0.2)

<14>2015 Jun 24 22: -admin-shell: %WAAS-PARSER-6-350232: CLI_LOG log_cli_command: show flash ", "@version"=>"1", "@timestamp"=>"2015-06-24T22:04:39.372Z", "type"=>"cisco", "host"=>"xxx.xxx.xxx.xxx"}, "log-type"]}>>]]}, :batch_timeout=>1, :force=>true, :final=>nil, :level=>:debug, :file=>"stud/buffer.rb", :line=>"207", :method=>"buffer_flush"}

**Comment by suyograo** [comment]
@dirtdiver512 can you please upgrade to 1.5.1 and try with your data? We fixed a bunch of issues with LS hanging in 1.5.1. Let us know how it goes!

**Comment by dirtdiver512** [comment]
Done....thanks and will let everyone know if the issue comes back.  should know in the next 24 hours

**Comment by KlavsKlavsen** [comment]
it's been stable after downgrade to 1.4.2.. I hope to find time monday, to upgrade 1 of the 4, to 1.5.1..

**Comment by dirtdiver512** [comment]
Well, no luck system still stopped.  No think in the logs, and can not stop the process vi service logstash stop.   I have to kill it with a -9 then start it.

**Comment by colinsurprenant** [comment]
maybe related to #3451 ?

**Comment by KlavsKlavsen** [comment]
very likely #3451 is the reason they won't shutdown. but why it kills redis with 1.5.x and 1.4.2 does not.. I haven't had time to do some tcpdumping to see what happens differently yet :(

**Comment by colinsurprenant** [comment]
@KlavsKlavsen it kills Redis?! can you provide more information on this? is Redis just unresponsive? maybe reached its memory limit? I never heard of a situation - other than Redis OOM - where interactions between Redis and Logstash are causing Redis hangs? but it could be something new...

**Comment by dirtdiver512** [comment]
All,  

I upgraded to the latest version and so far no crashed for over a week......here's crossing the fingers

**Comment by colinsurprenant** [comment]
I added this issue in the #3451 meta issue list.

**Comment by KlavsKlavsen** [comment]
by kills I mean it seems to take every connections it can handle so it stops accepting new connections.. its not OOM'ing. I've increated max connections to 100k - and yet it did the same.
reverting to 1.4.2 it didn't do that. I'll try upgrade again after vacation (I'm on it currently :) - and be sure to have connection counters on redis, so I can get some stats.

**Comment by nikut** [comment]
Just upgraded my "processing" instances (that pull logs from redis buffers and feed it to Elasticsearch) from 1.4.2 to 1.5.3 yesterday and seeing the same: they stop processing logs after 30-45 minutes and seem to eventually eat up massive numbers of redis connections. Had to downgrade to 1.4.2 for now.

So far 1.5.3 seems to be stable running as an agent reading log files and shipping to redis, though.

**Comment by jamconner** [comment]
Having similar issue with logstash 1.5.3 on Windows machine.

Using logstash to ship logs to elasticsearch cluster (1.7.1) on linux VMs, and after some period of time logstash 'hangs' on the windows machine.  It doesn't completely stop indexing logs, but the rate at which it indexes them slows down to almost nothing.  The process doesn't die, but CPU for the Windows JRuby Launcher just sits at 13% and memory stays flat.  Haven't been able to identify the trigger for the hang.  Sometimes logstash will run for many hours before hanging, and sometimes its much more frequent.  A simple restart of the process kicks back into action - and downgrade to logstash 1.4.2 seems to solve the hanging problem.  However i can't use 1.4.2 because it seems to put a lock on my log file and won't let the application roll it properly so app logs grow to a huge size.

So back to logstash_1.5.3 - when running from the cmd line, I don't see any relevant output that would suggest why the indexing grinds to a halt.  Could this still be related to the output buffer flush issue that was solved in this latest version?  Any thoughts are certainly appreciated.

**Comment by itsysadmins** [comment]
We've started seeing this issue with 'ELK'.  Using 1.7.1 of elasticsearch and 1.5.4 of logstash.  For us, appears it's an unknown number of entries per time period.  As long as it's a lower number it runs smooth but having very low to high range when it starts happening and can't say for sure.  *Appears* to be between 4K and 8K per minute where it starts and the larger the number of trx the more frequent it occurs.  We rolled back to 1.4.x for now since this issue seems to have occurred with 1.5 beta/rc and no solution yet.
